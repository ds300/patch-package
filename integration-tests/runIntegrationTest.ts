import * as fs from "fs-extra"
import { join, resolve } from "../src/path"
import * as tmp from "tmp"
import { spawnSafeSync } from "../src/spawnSafe"
import { resolveRelativeFileDependencies } from "../src/resolveRelativeFileDependencies"

export const patchPackageTarballPath = resolve(
  fs
    .readdirSync(".")
    .filter((nm) => nm.match(/^patch-package\.test\.\d+\.tgz$/))[0],
)

export function runIntegrationTest({
  projectName,
  shouldProduceSnapshots,
}: {
  projectName: string
  shouldProduceSnapshots: boolean
}) {
  describe(`Test ${projectName}:`, () => {
    const tmpDir = tmp.dirSync({ unsafeCleanup: true })
    fs.copySync(join(__dirname, projectName), tmpDir.name, {
      recursive: true,
    })

    const packageJson = require(join(tmpDir.name, "package.json"))
    packageJson.dependencies = resolveRelativeFileDependencies(
      join(__dirname, projectName),
      packageJson.dependencies,
    )

    fs.writeFileSync(
      join(tmpDir.name, "package.json"),
      JSON.stringify(packageJson),
    )

    const result = spawnSafeSync(
      `./${projectName}.sh`,
      [patchPackageTarballPath],
      {
        cwd: tmpDir.name,
        throwOnError: false,
      },
    )

    it("should exit with 0 status", () => {
      expect(result.status).toBe(0)
    })

    const output = result.stdout.toString() + "\n" + result.stderr.toString()

    if (result.status !== 0) {
      console.error(output)
    }

    it("should produce output", () => {
      expect(output.trim()).toBeTruthy()
    })

    const snapshots = output.match(/SNAPSHOT: ?([\s\S]*?)END SNAPSHOT/g)

    if (shouldProduceSnapshots) {
      it("should produce some snapshots", () => {
        expect(snapshots && snapshots.length).toBeTruthy()
      })
      if (snapshots) {
        snapshots.forEach((snapshot) => {
          const snapshotDescriptionMatch = snapshot.match(/SNAPSHOT: (.*)/)
          if (snapshotDescriptionMatch) {
            it(snapshotDescriptionMatch[1], () => {
              expect(snapshot).toMatchSnapshot()
            })
          } else {
            throw new Error("bad snapshot format")
          }
        })
      }
    } else {
      it("should not produce any snapshots", () => {
        expect(snapshots && snapshots.length).toBeFalsy()
      })
    }
  })
}
