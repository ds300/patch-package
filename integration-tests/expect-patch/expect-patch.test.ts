import { runIntegrationTest } from "../runIntegrationTest"
runIntegrationTest({
  projectName: "expect-patch",
  shouldProduceSnapshots: true,
})
