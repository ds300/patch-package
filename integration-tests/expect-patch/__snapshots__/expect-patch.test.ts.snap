// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test expect-patch: 00: expect-patch flag prints an error message and by default does not fail 1`] = `
"SNAPSHOT: expect-patch flag prints an error message and by default does not fail
patch-package 0.0.0
Applying patches...
No patch files found
---
patch-package finished with 1 error(s).
END SNAPSHOT"
`;

exports[`Test expect-patch: 01: expect-patch is silent when patches happen 1`] = `
"SNAPSHOT: expect-patch is silent when patches happen
patch-package 0.0.0
Applying patches...
left-pad@1.3.0 âœ”
END SNAPSHOT"
`;

exports[`Test expect-patch: 02: expect-patch flag produces error for error-on-fail flag 1`] = `
"SNAPSHOT: expect-patch flag produces error for error-on-fail flag
patch-package 0.0.0
Applying patches...
No patch files found
---
patch-package finished with 1 error(s).
END SNAPSHOT"
`;
